### ISSUE TITLE
Session loss: No batchPartInfo could found for the current sessionId

### ISSUE THREAD

**Issue by BluePalmTree** [Mimetis/Dotmim.Sync#735]
Hi,

if we have a long sync (~40min) at the end we always get this exception. What could this be?

Error during Config sync

Dotmim.Sync.Web.Client.HttpSyncWebException: Session loss: No batchPartInfo could found for the current sessionId. It seems the session was lost. Please try again.
-----------------------

-----------------------
-----------------------
INNER EXCEPTION
-----------------------
Session loss: No batchPartInfo could found for the current sessionId. It seems the session was lost. Please try again.
-----------------------
   at Dotmim.Sync.Web.Server.WebServerOrchestrator.HandleRequestAsync(HttpContext httpContext, Action`1 action, CancellationToken cancellationToken, IProgress`1 progress)
-----------------------

   at Dotmim.Sync.Web.Client.HttpRequestHandler.<HandleSyncError>d__15.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Dotmim.Sync.Web.Client.HttpRequestHandler.<SendAsync>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Dotmim.Sync.SyncPolicy.<InternalExecuteAsync>d__17`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Dotmim.Sync.Web.Client.HttpRequestHandler.<ProcessRequestAsync>d__12.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Dotmim.Sync.Web.Client.WebClientOrchestrator.<ApplyThenGetChangesAsync>d__36.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.ValidateEnd(Task task)
   at Dotmim.Sync.SyncAgent.<>c__DisplayClass53_0.<<SynchronizeAsync>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Dotmim.Sync.SyncAgent.<SynchronizeAsync>d__53.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   at Geeks.Pos.DataModel.PosDBSync.<SyncConfigAsync>d__16.MoveNext()

**Comment by Mimetis** [comment]
your server lost your session, for some reason

**Comment by matevzcesar** [comment]
@BluePalmTree is 40 minutes the usual sync duration for you? I'm asking because we are also seeing logs of this error, and sometimes (SyncType.Normal) synchronizations connected to this error are taking longer than usual (10 to 20 min). 

This can happen multiple times in a row, but users have reported that if they run a Reinitializing sync, it usually solves the problem and finishes in a few minutes (2-5 min). What's also strange is that when we see that a synchronization is taking longer than usual, if we check the database for changes, all the changes from the client are already applied but the sync seems to be "stuck". And it does not make sense to me why a SyncType.Reinitialize sync would fix this.

Are you experiencing anything similar? I'm just trying to understand if these symptoms have anything to do with this error. They really shouldn't, but this is my only lead.

Also which version of the lib. are you using? I have noticed that in previous versions there were logs of another error which I [have reported here](https://github.com/Mimetis/Dotmim.Sync/issues/719), but since upgrading to 0.9.4 the error you posted seems to have "replaced" it.

**Comment by gentledepp** [comment]
Hi! This exception was actually introduced by me as I found an issue in Dotmim.Sync.

When uploading changes in batches, the client sends each batch part in a separate request.
On the server-side, all those batches are stored on disk
only when the *last batch file* is uploaded, all those files are deserialized and the changes they contain are applied to the database in one transaction.
To know which files belong together, DMS stores a BatchPartInfo in the session - which is normally just an in-memory cache.

So _should_ this session be cleared (e.g. IIS decides to recycle an application pool which can happen once in a while), this `batchPartInfo` is suddenly lost.

This is actually a good thing as it will cause the client to re-send all batches.

If we did not throw this exception, all previous batch files would _never_ be applied to the database. So you would in fact lose changes and not even know about it. This was really a critical issue for us and our customers.

**Comment by matevzcesar** [comment]
@gentledepp Thanks for the insight.
Do you have an idea what could be causing a session to be cleared aside from application pool recycle? I'm looking at the application pool recycle logs and they do not match up with reported errors. We get approx. 5-10 logs of the error per day from different users, the last logged application pool recycle was two weeks ago. 

Can you maybe share how frequently you experience this error?

**Comment by BluePalmTree** [comment]
Hello, sorry I haven't looked into this issue to deeply because even if this error everything is still synced. So on the next sync everything is already in the database.
We run the sync on every start of the app, and this error onyl occurs on the initial sync of the database. And only if we have large syncs (~40min with ~ 9 million rows). We use version 0.9.1.

**Comment by matevzcesar** [comment]
No worries, thanks for the info! Seems we have different issues.

**Comment by gentledepp** [comment]
@matevzcesar I am not entirely sure.
First, please check what kind of session implementation you use.

It could also be, that you have a scaleout scenario (two or more web servers) and you did not configure a distributed cache like redis. If the first batch is then transmitted to server 1, but the second to server 2 you would have the same issue.

Or, your session is kept for e.g. 5 minutes (in the cache), but you sync too much data. => session is cleared, and you lose your batchPartInfo

In other words: I just do not know enough about your setup to tell you what is going on.

**Comment by matevzcesar** [comment]
@gentledepp Thank you for the tips! I'll investigate further and report back any findings.
